### PR TITLE
Feature/animate layer transition

### LINKED ISSUE #26: Animate Layer Transitions
When the artist's base layer is becoming inactive it needs to be visually animated. This should be a general approach where future animations could be added to layers.

### PR DESCRIPTION
**Pull request by mpryor**
This PR closes https://github.com/jollyjerr/mindgraph/issues/26

## What does this PR do?
This PR introduces the ability to "transition" nodes between layers, and attach animations to layers during those transitions.

## Road blocks
So honestly I hit a TON of obstacles while working on this.

### Road block 1 - the need for a transition or delay
Initially, I thought this feature would be trivial, and my first [attempt](https://github.com/jollyjerr/mindgraph/commit/207f791ec8fa8de9a305a28166c60fd5570c69e3) was like 90% there, by simply attaching an animation property to each layer, having those layers animate opacity when certain events happen, and calling it a day. This solution works great when transitioning a node from the base layer to the active layer, because the node immediately joins the active layer and then the animation of transitioning between "neutral" and "active" state is handled by the Drawable itself. 

This unfortunately did not really work for when a node leaves the active layer and joins the base layer again, because the base layer is animating from dimmed to full opacity again. If the active node immediately joins that layer, it goes dim and it looks pretty wonky. This led to the concept of a `LayerTransition`. 

### Layer Transitions
Essentially, this concept introduces the ability for a Drawable to leave the layer it is on, hanging out in layer purgatory for some set duration, and then joining the layer it's trying to go to. A LayerTransition has a lot of the same properties as a normal Layer and is even drawn alongside the regular layers. 

The introduction of this concept came with a lot of baggage and some new issues to solve...

### Road block 2 - Layer logic
The logic for transitioning between layers, animating the transition, etc ended up being a lot of code and created the need for ANOTHER class, the `TransitionManager`. This thing just wraps a lot of stateful logic for managing layers, animations, and layer transitions. 

### Road block 3 - Handling re-triggers
After introducing LayerTransitions, I quickly ran into the issue of re-triggering a transition in the middle of a running transition. Getting animations to cleanly play again in certain scenarios ended up being kinda tricky... for example: move your mouse onto a node, leave the node, and immediately re-enter it. 

### Road block 4 - zIndex
Since LayerTransitions are basically layers themselves, I started running into issues with the rendering order of nodes. When drawing nodes on a LayerTransition, the link elements were rendering on top of the active/base layer during transitioning, and then `sort`'ing itself out after the transition ended. This wasn't too big of a deal, but I didn't like it so I wrote some code to fix it... and that might be some of the ugliest code I've written on this project so far [this monstrosity](https://github.com/jollyjerr/mindgraph/commit/7c2b2175122ba9f07e4f7c2f05d1042d3ea4f20f#diff-1b6ad78e77b4b2649a857d61600007964174ba5ee1a6d0b69517fd3b9c27e151R113)

I'm going to take a break from this one for a bit and revisit it and clean it up. If you review this and just have `nits`, then maybe I'll clean it up the future instead. Yikes!

### CODE REVIEW

File: packages/draw/src/artist.ts [hunk 1]
```diff
@@ -2,17 +2,42 @@ import { Canvas, Drawable } from './canvas';
...
   drawables: Drawable[];
   focus: Focus;
+  animation?: Animation<number>;
 }
 
+const DIMMING_ANIMATION_CONFIG: (styles: Styles) => AnimationConfig<number> = (
+  styles: Styles,
+) => {
+  return {
+    duration: 1,
```
Comment (mpryor) [inline packages/draw/src/artist.ts:23]: I probably need to externalize the duration of our animations and transitions to `styles`

Comment (mpryor) [reply packages/draw/src/artist.ts]: I ended up doing this

**Comment by mpryor** [comment]
@jollyjerr I think this thing is ready... tear it up

**Review by jollyjerr — COMMENTED** [review]
Bro this is epic. My review is in https://github.com/jollyjerr/mindgraph/pull/31

**Review by jollyjerr — APPROVED** [review]

**Pull request merged**